rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,757 @@
+                # combining layer
+                new_signature = FeatureSignature(
+                    is_sparse=False,
+                    units=self._tf_layers[f"sparse_dense.{feature_type}"].output_units,
+                    number_of_dimensions=original_signature.number_of_dimensions,
+                )
+                new_signatures[feature_type] = new_signature
+
+            # prepare the combining layer itself
+            self._tf_layers["concat_seq_sent"] = ConcatenateSequenceSentenceFeatures(
+                attribute=attribute,
+                feature_signature_sequence=new_signatures[SEQUENCE],
+                feature_signature_sentence=new_signatures[SENTENCE],
+                concat_dimension=config[CONCAT_DIMENSION].get(attribute, None),
+                config=config,
+            )
+
+        # if only one feature type is present, prepare for only combining sparse & dense
+        # features
+        if self.have_all_feature_types:
+            self.output_units = self._tf_layers["concat_seq_sent"]
+        elif f"sparse_dense.{SEQUENCE}" in self._tf_layers:
+            self.output_units = self._tf_layers[f"sparse_dense.{SEQUENCE}"].output_units
+        else:
+            self.output_units = self._tf_layers[f"sparse_dense.{SENTENCE}"].output_units
+
+    def call(
+        self,
+        inputs: Tuple[
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            tf.Tensor,
+            tf.Tensor,
+        ],
+        training: Optional[Union[tf.Tensor, bool]] = None,
+    ) -> tf.Tensor:
+        """Combine multiple 3-D dense/sparse feature tensors into one.
+
+        Arguments:
+            inputs: Tuple containing:
+                sequence_features: List of 3-D dense or sparse tensors with token-level
+                    features.
+                sentence_features: List of 3-D dense or sparse tensors with sentence-level
+                    features.
+                mask_sequence: a 3-D tensor mask that has 1s at real and 0s at padded
+                    positions corresponding to tokens in `sequence_features`.
+                mask_sequence_sentence: a 3-D tensor mask similar to `mask_sequence` but
+                    having each sequence of 1s longer by 1 to account for sequence 
+                    lengths of sequence- and sentence-level features being combined.
+            training: Python boolean indicating whether the layer should behave in
+                training mode (applying dropout to sparse tensors if applicable) or in
+                inference mode (not applying dropout).
+
+        Returns:
+            Single dense 3-D tensor containing the combined sequence- and sentence-
+            level, sparse & dense features.
+        """
+        sequence_features = inputs[0]
+        sentence_features = inputs[1]
+        mask_sequence = inputs[2]
+        mask_sequence_sentence = inputs[3]
```

Comment:
this is a weird name 😅, what do you mean?

Reply:
You mean `mask_sequence_sentence`? My thinking is: `mask_sequence` masks sequence-level features. The 2nd mask masks the combined sequence- and sentence-level features. Hence the name. Previously, the naming was quite chaotic, sometimes using `mask` or `mask_text`, but I'm not a fan of `mask_text` as it is confusing given the `TEXT` attribute...

Reply:
Wait, there was even `mask_sequence_text` previously, which was the hardest one to understand 😛 My new name is quite long but if you read the rest of the code and the docstrings, I think it's quite consistent...

Reply:
I'd make it even longer with `mask_combined_sequence_sentence`, it's very long but at least it is more concrete

Reply:
Excellent, I love long self-explanatory names 😄